src/main/java/com/uid2/admin/secret/SaltRotation.java

```diff
@@ -3,58 +3,89 @@
         this.keyGenerator = keyGenerator;
         this.isRefreshFromEnabled = config.getBoolean(AdminConst.ENABLE_SALT_ROTATION_REFRESH_FROM, false);
     }
 
-    public Result rotateSalts(RotatingSaltProvider.SaltSnapshot lastSnapshot,
-                                                        Duration[] minAges,
-                                                        double fraction,
-                                                        LocalDate targetDate) throws Exception {
-        final Instant nextEffective = targetDate.atStartOfDay().toInstant(ZoneOffset.UTC);
-        final Instant nextExpires = nextEffective.plus(7, ChronoUnit.DAYS);
+    public Result rotateSalts(
+            SaltSnapshot lastSnapshot,
+            Duration[] minAges,
+            double fraction,
+            LocalDate targetDate
+    ) throws Exception {
+        var preRotationSalts = lastSnapshot.getAllRotatingSalts();
+        var nextEffective = targetDate.atStartOfDay().toInstant(ZoneOffset.UTC);
+        var nextExpires = nextEffective.plus(7, ChronoUnit.DAYS);
         if (nextEffective.equals(lastSnapshot.getEffective()) || nextEffective.isBefore(lastSnapshot.getEffective())) {
             return Result.noSnapshot("cannot create a new salt snapshot with effective timestamp equal or prior to that of an existing snapshot");
         }
 
-        List<Integer> saltIndexesToRotate = pickSaltIndexesToRotate(lastSnapshot, nextEffective, minAges, fraction);
+        var rotatableSaltIndexes = findRotatableSaltIndexes(preRotationSalts, nextEffective.toEpochMilli());
+        var saltIndexesToRotate = pickSaltIndexesToRotate(
+                nextEffective,
+                minAges,
+                fraction,
+                preRotationSalts,
+                rotatableSaltIndexes
+        );
+
         if (saltIndexesToRotate.isEmpty()) {
-            return Result.noSnapshot("all salts are below min rotation age");
+            return Result.noSnapshot("all rotatable salts are below min rotation age");
         }
 
-        var updatedSalts = updateSalts(lastSnapshot.getAllRotatingSalts(), saltIndexesToRotate, nextEffective.toEpochMilli());
+        var postRotationSalts = updateSalts(preRotationSalts, saltIndexesToRotate, nextEffective.toEpochMilli());
 
-        SaltSnapshot nextSnapshot = new SaltSnapshot(
+        logSaltCounts(
+                targetDate,
+                nextEffective,
+                preRotationSalts,
+                postRotationSalts,
+                rotatableSaltIndexes,
+                saltIndexesToRotate
+        );
+
+        var nextSnapshot = new SaltSnapshot(
                 nextEffective,
                 nextExpires,
-                updatedSalts,
+                postRotationSalts,
                 lastSnapshot.getFirstLevelSalt());
         return Result.fromSnapshot(nextSnapshot);
     }
 
+    private List<Integer> findRotatableSaltIndexes(SaltEntry[] preRotationSalts, long nextEffective) {
```

Comment:
Is there a reason why we can't use salt bucket IDs rather than indexes?

Reply:
could even just pass around arrays of salt entries and override its equal method if that works better

Reply:
We could, I wanted to be consistent with how `saltIndexesToRotate` is done, which is indexes. 

I guess that's done for performance reasons, so we can address specific indexes in the array of >1m rather than cycling through it. I can look at how much difference switching to objects across the board would make, might not be enough to matter.